### PR TITLE
Boost colors support

### PR DESCRIPTION
**Pull request by JonnyVR1**

### CODE REVIEW

File: ExtendedColorSchemes/Directory.Build.targets [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+      <!--If Beat Saber is running, output to the Pending folder-->
+      <PluginDir Condition="'$(IsRunning)' == 'True'">$(BeatSaberDir)\IPA\Pending\Plugins</PluginDir>
+    </PropertyGroup>
+    <Warning Text="$(CopyToPluginsError)" Condition="'$(CopyToPluginsError)' != ''" />
+    <Message Text="Copying '$(OutputAssemblyName).dll' to '$(PluginDir)'." Importance="high" Condition="$(CanCopyToPlugins)" />
+    <Copy SourceFiles="$(OutputAssemblyName).dll" DestinationFiles="$(PluginDir)\$(AssemblyName).dll" Condition="$(CanCopyToPlugins)"  />
+    <Copy SourceFiles="$(OutputAssemblyName).pdb" DestinationFiles="$(PluginDir)\$(AssemblyName).pdb" Condition="'$(CanCopyToPlugins)' == 'True' AND Exists('$(OutputAssemblyName).pdb')"  />
+    <Warning Text="Beat Saber is running, restart the game to use the latest build." Condition="'$(IsRunning)' == 'True'" />
+  </Target>
+</Project>
```
Comment (Meivyn) [inline ExtendedColorSchemes/Directory.Build.targets]: This entire file is not needed anymore with BSMT.T beta, please remove it.

File: ExtendedColorSchemes/HarmonyPatches/PlayerDataFileManagerSO.cs [hunk 2]
```diff
@@ -40,6 +40,14 @@ private static void Postfix()
...
                 File.Copy(filePath, backupFilePath, true);
+					
+                // Only keep the last 5 days worth of files to stop file spam.
+                var fiveDaysAgo = DateTime.Now.Date.AddDays(-5);
+                var oldFiles = Directory.EnumerateFiles(Application.persistentDataPath, "PlayerData_*", SearchOption.TopDirectoryOnly)
+                    .Where(path => File.GetCreationTime(path) > fiveDaysAgo)
+                    .ToList();
+					
+                oldFiles.ForEach(File.Delete);
             }
```
Comment (Meivyn) [inline ExtendedColorSchemes/HarmonyPatches/PlayerDataFileManagerSO.cs]: While I think this change is a good thing, I'd appreciate if you could make a separate commit for that.

File: ExtendedColorSchemes/Converters/ColorSchemeWithEditableNameConverter.cs [hunk 3]
```diff
@@ -34,7 +34,10 @@ internal class ColorSchemeWithEditableNameConverter : ValueConverter<ColorScheme
                 CustomValueTypeConverter<Color>.Deserialize(map["saberBColor"], parent),
                 CustomValueTypeConverter<Color>.Deserialize(map["environmentColor0"], parent),
                 CustomValueTypeConverter<Color>.Deserialize(map["environmentColor1"], parent),
-                CustomValueTypeConverter<Color>.Deserialize(map["obstaclesColor"], parent));
+                CustomValueTypeConverter<Color>.Deserialize(map["obstaclesColor"], parent),
+                bool.Parse(map["supportsEnvironmentColorBoost"]?.ToString() ?? "true"),
```
Comment (Meivyn) [inline ExtendedColorSchemes/Converters/ColorSchemeWithEditableNameConverter.cs]: This doesn't make use of the `Value` type of BSIPA, you should rather make something like:

```cs
// Make sure this uses IPA.Config.Data.Boolean and not System.Boolean
if (map["supportsEnvironmentColorBoost"] is not Boolean supportsEnvironmentColorBoost)
{
    throw new ArgumentException("supportsEnvironmentColorBoost must be a boolean!");
}
```

And then access the value directly:

```cs
supportsEnvironmentColorBoost.Value
```

File: ExtendedColorSchemes/Converters/ColorSchemeWithEditableNameConverter.cs [hunk 4]
```diff
@@ -52,7 +55,10 @@ internal class ColorSchemeWithEditableNameConverter : ValueConverter<ColorScheme
                 {"saberBColor", CustomValueTypeConverter<Color>.Serialize(obj.saberBColor)},
                 {"environmentColor0", CustomValueTypeConverter<Color>.Serialize(obj.environmentColor0)},
                 {"environmentColor1", CustomValueTypeConverter<Color>.Serialize(obj.environmentColor1)},
-                {"obstaclesColor", CustomValueTypeConverter<Color>.Serialize(obj.obstaclesColor)}
+                {"obstaclesColor", CustomValueTypeConverter<Color>.Serialize(obj.obstaclesColor)},
+                {"supportsEnvironmentColorBoost", Value.Bool(obj.supportsEnvironmentColorBoost)},
```
Comment (Meivyn) [inline ExtendedColorSchemes/Converters/ColorSchemeWithEditableNameConverter.cs]: I'd use `Value.From` there.

File: ExtendedColorSchemes/HarmonyPatches/ColorSchemesSettings.cs [hunk 5]
```diff
@@ -25,6 +25,9 @@ internal static void Prefix(ref ColorScheme[] colorSchemes)
 
             List<ColorScheme> colorSchemesList = colorSchemes.ToList();
 
+            var first = colorSchemesList.First();
+            var firstWithBoostColors = colorSchemesList.FirstOrDefault(x => x.supportsEnvironmentColorBoost);
```
Comment (Meivyn) [inline ExtendedColorSchemes/HarmonyPatches/ColorSchemesSettings.cs:29]: For now this is okay, but you should be fine to just use `first` as well, currently it might be needed because I have missed a patch in the `PlayerDataFileManagerSO`.

**Review by Meivyn — CHANGES_REQUESTED** [review]

**Comment by Meivyn** [comment]
Thanks for your PR. It shouldn't be needed to be honest, but devs once again rushed this and completely forgot to make the boost colors savable.

**Comment by JonnyVR1** [comment]
> Thanks for your PR. It shouldn't be needed to be honest, but devs once again rushed this and completely forgot to make the boost colors savable.

Yeah the base game just uses color0 and color1 as boost colors which is just weird. I'll drop Jan a message to escalate this

File: ExtendedColorSchemes/HarmonyPatches/ColorSchemesSettings.cs (hunk 5 above)
Comment (Meivyn) [reply ExtendedColorSchemes/HarmonyPatches/ColorSchemesSettings.cs:29]: As I thought, this doesn't seems to do a thing due to the missing patch.

image

File: ExtendedColorSchemes/HarmonyPatches/PlayerDataFileManagerSO.cs [hunk 6]
```diff
@@ -56,9 +56,17 @@ private static void GetColorSchemes(ColorSchemesSettings colorSchemesSettings, r
...
+                        savedColorScheme?.colorSchemeName ?? string.Empty,
+                        savedColorScheme?.colorSchemeId ?? colorScheme.colorSchemeId,
+                        savedColorScheme?.saberAColor ?? colorScheme.saberAColor,
+                        savedColorScheme?.saberBColor ?? colorScheme.saberBColor,
+                        savedColorScheme?.environmentColor0 ?? colorScheme.environmentColor0,
+                        savedColorScheme?.environmentColor1 ?? colorScheme.environmentColor1,
+                        savedColorScheme?.obstaclesColor ?? colorScheme.obstaclesColor,
+                        savedColorScheme?.supportsEnvironmentColorBoost ?? colorScheme.supportsEnvironmentColorBoost,
+                        savedColorScheme?.environmentColor0Boost ?? colorScheme.environmentColor0Boost,
+                        savedColorScheme?.environmentColor1Boost ?? colorScheme.environmentColor1Boost);
```
Comment (Meivyn) [inline ExtendedColorSchemes/HarmonyPatches/PlayerDataFileManagerSO.cs]: That stuff simply doesn't save anything. You should keep the reference to `colorScheme` because the color schemes aren't saved by that point. The whole point of `savedColorScheme` here is to grab manual edits of the json (for the name change that you can only do by editing it manually), using `colorScheme` should be enough to save everything else.

**Review by Meivyn — APPROVED** [review]

**Review by Meivyn — CHANGES_REQUESTED** [review]
Should learn how to use that review thing...

**Review by Meivyn — APPROVED** [review]
